server/test/com/cloud/vm/UserVmManagerTest.java

```diff
@@ -928,4 +928,13 @@ public void testUpdateVmNicIpFailure3() throws Exception {
             CallContext.unregister();
         }
     }
+
+    @Test
+    public void testPersistDeviceBusInfo() {
+        when(_vmMock.getDetail(any(String.class))).thenReturn(null);
+        _userVmMgr.persistDeviceBusInfo(_vmMock, null);
+        verify(_vmDao, times(0)).saveDetails(any(UserVmVO.class));
+        _userVmMgr.persistDeviceBusInfo(_vmMock, "lsilogic");
+        verify(_vmDao, times(1)).saveDetails(any(UserVmVO.class));
```

Comment:
Why not split these two ``verify`` into two separate test methods to better express the scenario being tested?  It would help with diagnosing the specific failure and make the code more maintainable.

Also, please add test cases for when the ``existingVmRootDiskController`` is empty and ``rootDiskController`` is empty.